src/toil/jobStores/aws/jobStore.py

```diff
@@ -1047,3 +1011,75 @@ def toItem(self):
+    """
+    Copies a key from a source key to a destination key in multiple parts. Note that if the
+    destination key exists it will be overwritten implicitly, and if it does not exist a new
+    key will be created. If the destination bucket does not exist an error will be raised.
+
+    :param boto.s3.key.Key srcKey: The source key to be copied from.
+    :param str dstBucketName: The name of the destination bucket for the copy.
+    :param str dstKeyName: The name of the destination key that will be created or overwritten.
+    :param dict headers: Any headers that should be passed.
+
+    :rtype: boto.s3.multipart.CompletedMultiPartUpload
+    :return: An object representing the completed upload.
+    """
+    s3 = boto.connect_s3()
+    partSize = defaultPartSize
+    totalSize = srcKey.size
+    totalParts = (totalSize / partSize) + 1 if totalSize % partSize else (totalSize / partSize)
+    exceptionList = []
+
+    def _copyPart(partIndex):
+        if exceptionList:
+            return None
+        else:
+            try:
+                startByte = partIndex * partSize
+                startByte = startByte if startByte <= totalSize else totalSize - ((partIndex - 1) * partSize)
+
+                part = upload.copy_part_from_key(src_bucket_name=srcKey.bucket.name,
+                                                 src_key_name=srcKey.name,
+                                                 part_num=partIndex + 1,  # Part numbers are indexed from 1 in AWS.
+                                                 start=startByte,
+                                                 end=min(startByte + partSize, totalSize) - 1,
+                                                 headers=headers or {})
+
+                log.debug('Copied part %d of %d' % (partIndex, totalParts))
+
+            except:
+                exceptionList.append({'partIndex': partIndex, 'tb': traceback.format_exc()})
+                log.debug('Copy failed.')
+                return None
+            else:
+                return part
+
+    upload = s3.get_bucket(dstBucketName).initiate_multipart_upload(dstKeyName, headers=headers)
+    log.info("Initiated multipart copy from 's3://%s/%s' to 's3://%s/%s'." %
+             (srcKey.bucket.name, srcKey.name, dstBucketName, dstKeyName))
+    try:
+        with ThreadPoolExecutor(max_workers=min(cpu_count() * 16, totalParts, 128)) as executor:
+            finishedParts = set(executor.map(_copyPart, range(0, totalParts)))
+
+            if exceptionList:
+                i = 0
+                for exceptionInfo in exceptionList:
+                    if i >= 4:  # We log a maximum of four exceptions to prevent over logging.
+                        break
+                    else:
+                        log.info("Part %d of multipart copy failed with the following exception:\n%s" %
+                                 (exceptionInfo['partIndex'], exceptionInfo['tb']))
+                        i += 1
```

Comment:
Don't you also need to raise an exception as the last element in the if body, such that cancel_upload is triggered?